In [1]:
import numpy as np
import pandas as pd

In [47]:
data = pd.read_csv("2023_ai_inventory.csv")

    #data['agency_edit'] = data['Agency'].fillna(data['Department'])


In [73]:
data['Development_Stage'].fillna('Deployment').unique()

array(['Initiation', 'Development and Acquisition',
       'Operation and Maintenance', 'Deployment', 'Implementation',
       'Planned (not in production)', 'Completed',
       'Development (not in production)',
       'Investigating/Proof of concept', 'Proof-of-concept completed',
       'In production: more than 1 year',
       'In production: less than 6 months',
       'Technical transfer of capability to industry planned this summer.',
       'Successfully tested but not in production.',
       'Refinments planned for future release', 'Initial development',
       "User Acceptance Testing to begin early spring '22",
       'Proof of Concept completed and published',
       'Currently under development',
       'OPC runs in a pseudo-operational capacity via a webpage maintained by the Massachusetts Institute of Technology - Lincoln Lab, as well as in a test and evaluation capacity in a research mode.',
       'This is a pilot initiative', 'In-use', 'In mission',
       'In product

In [75]:
def create_dev_position(df):
    """
    Create a new column for the department x y for chart.
    """

    dev_x_dict = {}
    # for now
    df['Development_Stage'].fillna('Deployment', inplace=True)
    
    devs = df['Development_Stage'].unique().tolist()
    devs.sort()
    # 5 wide means first is at 3 then +5
    i = 0
    for dev in devs:
        dev_x_dict[dev] = i
        i += 1

    df['dev_rn'] = df.sort_values(['Development_Stage','Title'], ascending=[True,True]) \
             .groupby(['Development_Stage']) \
             .cumcount()
    df['dev_x'] = (df['Development_Stage'].map(dev_x_dict) * 7) + (df['dev_rn'] % 5) +1
    df['dev_y'] = df['dev_rn'] // 5 + 1

    return df
    

In [61]:
def create_dept_position(df):
    """
    Create a new column for the department x y for chart.
    """

    dept_x_dict = {}
    depts = df['Department_Code'].unique().tolist()
    
    depts.sort()
    # 5 wide means first is at 3 then +5
    i = 0
    for dept in depts:
        dept_x_dict[dept] = i
        i += 1

    df['dept_rn'] = df.sort_values(['Department_Code','Title'], ascending=[True,True]) \
             .groupby(['Department_Code']) \
             .cumcount()
    df['dept_x'] = (df['Department_Code'].map(dept_x_dict) * 7) + (df['dept_rn'] % 5) +1
    df['dept_y'] = df['dept_rn'] // 5 + 1

    return df
    


In [76]:
data = create_dept_position(data)
data = create_dev_position(data)

In [77]:
data.sort_values(['Department_Code','Title'], ascending=True).head(10)

Use_Case_ID Department_Code  \
0   DHS-0000-2023             DHS   
15  DHS-0015-2023             DHS   
1   DHS-0001-2023             DHS   
16  DHS-0016-2023             DHS   
13  DHS-0013-2023             DHS   
14  DHS-0014-2023             DHS   
17  DHS-0017-2023             DHS   
2   DHS-0002-2023             DHS   
3   DHS-0003-2023             DHS   
4   DHS-0004-2023             DHS   

                                              Agency Office  \
0                      Customs and Border Protection    NaN   
15  Cybersecurity and Infrastructure Security Agency    NaN   
1                      Customs and Border Protection    NaN   
16  Cybersecurity and Infrastructure Security Agency    NaN   
13  Cybersecurity and Infrastructure Security Agency    NaN   
14  Cybersecurity and Infrastructure Security Agency    NaN   
17  Cybersecurity and Infrastructure Security Agency    NaN   
2                      Customs and Border Protection    NaN   
3                      Customs and Border Protection    NaN   
4                      Customs and Border Protection    NaN   

                                                Title  \
0                           AI Curated Synthetic Data   
15                         AI Security and Robustness   
1             AI for Autonomous Situational Awareness   
16                           AIS Scoring and Feedback   
13   Advanced Analytic Enabled Forensic Investigation   
14                  Advanced Network Anomaly Alerting   
17  Automated Indicator Sharing (AIS) Automated PI...   
2         Automated Item of Interest Detection - ICAD   
3                                 Autonomous Aerostat   
4                       Autonomous Maritime Awareness   

                                              Summary  \
0   AI Curated Synthetic Data creates synthetic da...   
15  Frameworks, processes, and testing tools devel...   
1   The AI for autonomous situational awareness sy...   
16  AIS Automated Scoring & Feedback (AS&F) is use...   
13  CISA deploys forensic specialists to analyze c...   
14  Threat hunting and Security Operations Center ...   
17  The Automated PII Detection and Human Review P...   
2   The software analyzes photographs that are tak...   
3   Aerostat capability that uses three tethers in...   
4   The Autonomous Maritime Awareness system combi...   

              Development_Stage  \
0                    Initiation   
15                   Initiation   
1   Development and Acquisition   
16    Operation and Maintenance   
13                   Initiation   
14                   Initiation   
17    Operation and Maintenance   
2     Operation and Maintenance   
3   Development and Acquisition   
4   Development and Acquisition   

                                           Techniques Source_Code  \
0                          Synthetic Image Generation         NaN   
15  Machine Learning, Natural Language Processing ...         NaN   
1                                      Machine Vision         NaN   
16        Descriptive Analysis; Machine Learning; NLP         NaN   
13                                   Machine Learning         NaN   
14                                   Machine Learning         NaN   
17                  Natural Language Processing (NLP)         NaN   
2                                    Machine Learning         NaN   
3                               Automation & Robotics         NaN   
4              Machine Learning Automation & Robotics         NaN   

                         Department  dept_rn  dept_x  dept_y  dev_rn  dev_x  \
0   Department of Homeland Security        0       1       1       1     93   
15  Department of Homeland Security        1       2       1       2     94   
1   Department of Homeland Security        2       3       1       0     29   
16  Department of Homeland Security        3       4       1       1    114   
13  Department of Homeland Security        4       5       1       4     96   
14  Department of Homeland S

In [78]:
data.to_csv("2023_ai_inventory_edit.csv", index=False)

In [5]:
# data clean up

# add a column for the agency filling NA for charts




In [6]:
data.head()


Use_Case_ID Department_Code                         Agency Office  \
0  DHS-0000-2023             DHS  Customs and Border Protection    NaN   
1  DHS-0001-2023             DHS  Customs and Border Protection    NaN   
2  DHS-0002-2023             DHS  Customs and Border Protection    NaN   
3  DHS-0003-2023             DHS  Customs and Border Protection    NaN   
4  DHS-0004-2023             DHS  Customs and Border Protection    NaN   

                                         Title  \
0                    AI Curated Synthetic Data   
1      AI for Autonomous Situational Awareness   
2  Automated Item of Interest Detection - ICAD   
3                          Autonomous Aerostat   
4                Autonomous Maritime Awareness   

                                             Summary  \
0  AI Curated Synthetic Data creates synthetic da...   
1  The AI for autonomous situational awareness sy...   
2  The software analyzes photographs that are tak...   
3  Aerostat capability that uses three tethers in...   
4  The Autonomous Maritime Awareness system combi...   

             Development_Stage                              Techniques  \
0                   Initiation              Synthetic Image Generation   
1  Development and Acquisition                          Machine Vision   
2    Operation and Maintenance                        Machine Learning   
3  Development and Acquisition                   Automation & Robotics   
4  Development and Acquisition  Machine Learning Automation & Robotics   

  Source_Code                       Department                    agency_edit  
0         NaN  Department of Homeland Security  Customs and Border Protection  
1         NaN  Department of Homeland Security  Customs and Border Protection  
2         NaN  Department of Homeland Security  Customs and Border Protection  
3         NaN  Department of Homeland Security  Customs and Border Protection  
4         NaN  Department of Homeland Security  Customs and Border Protection

In [6]:
data.describe()

Use_Case_ID Department_Code                                 Agency  \
count             710             710                                    634   
unique            710              21                                     85   
top     DHS-0000-2023             DOE  National Energy Technology Laboratory   
freq                1             178                                    127   

                                                Office  \
count                                              196   
unique                                              61   
top     Centers for Medicare & Medicaid Services (CMS)   
freq                                                23   

                                                Title  \
count                                             710   
unique                                            707   
top     Machine Learning for Occupant Safety Research   
freq                                                2   

                                                  Summary  \
count                                                 710   
unique                                                705   
top     Custom machine learning model to extract data ...   
freq                                                    2   

                Development_Stage Techniques  \
count                         389        315   
unique                         26        120   
top     Operation and Maintenance      Other   
freq                           91         31   

                               Source_Code            Department  
count                                   17                   710  
unique                                  17                    21  
top     https://github.com/usbr/PyForecast  Department of Energy  
freq                                     1                   178

There are 710 models
* every model has a unique case id and department code (should use department)
* 
* 10% are missing agency. adding new col for chart to just replace with department name
    * Departments either have agency or dont (HUD, Labor, Treasury, VA, and EPA have no agency info)
* Only Agriculture and Health and Human Services have office info
* 

In [27]:
(data.isnull().sum() / len(data)) * 100

Use_Case_ID           0.000000
Department_Code       0.000000
Agency               10.704225
Office               72.394366
Title                 0.000000
Summary               0.000000
Development_Stage    45.211268
Techniques           55.633803
Source_Code          97.605634
Department            0.000000
dtype: float64

In [41]:
pd.set_option('display.max_colwidth', None)
data[data.duplicated('Summary', keep=False)][["Title", "Summary"]]

# exact duplicates?
#these look to be different use cases, so we'll keep them all
#data[data['Use_Case_ID'].isin(['DOC-0024-2023', 'DOC-0025-2023'])]['Summary'].values

#data[data['Use_Case_ID'].isin(['DOT-0007-2023', 'DOT-0008-2023'])]['Summary'].values

#data[data['Use_Case_ID'].isin(['HHS-0006-2023', 'HHS-0007-2023'])]['Summary'].values

Title  \
140                             Fluid migration from well-to-well communication will be inputted in AI to determine a costs-benefit analysis   
159           To drive insights on the dependencies between the natural gas and electricity sectors to increase reliability of the NG system   
160  To drive insights on the power system reliability, cost, and operations during the energy transition with and without FECM technologies   
243       Using recursive neural networks and using fiber optic cables to recognize strain patterns and warn operators a fracture is coming.   
310                                                                                                       Form Recognizer for Benefits Forms   
316                                                                                                          Data Ingestion of Payroll Forms   
392                                                                                                                 R+2:18eDIRECT: Clarivate   
393                                                                                                                     ReDIRECT: AriScience   
696                                             Machine learning models to predict disease progression among veterans with hepatitis C virus   
701                                              Reinforcement learning evaluation of treatment policies for patients with hepatitis C virus   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Summary  
140  This project will develop an ML algorithm to predict the time when a \ngrowing fracture will reach the monitored well. The ML workflow will be \ntrained on the distinctive tensile strain signature that precedes the \ngrowing fracture. The new workflow will be designed to work in \nconjunction with the fracture warning ML workflow developed in EY21. \nTogether, these workflows will: (1) provide an early warning of well-to-\nwell communication, (2) predict the measured depths where the \ncommunication will happen, and (3) provide an estimated time until the \nbeginning of well-to-well communication.  
159                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Commercially available models will be used to generate predictive \nscenarios  
160                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Commercially available models will be used to generate predictive \nscenarios  
243  This project will develop an ML algorithm to predict the time when a \ngrowing fracture will reac

In [18]:


#which department has the most missing agency data
(data.groupby('Department')['Agency'].apply(lambda x: (x.isnull().sum() / len(x)) * 100))
(data.groupby('Department')['Office'].apply(lambda x: (x.isnull().sum() / len(x)) * 100))


Department
Department of Agriculture                          0.0
Department of Commerce                           100.0
Department of Education                          100.0
Department of Energy                             100.0
Department of Health and Human Services            0.0
Department of Homeland Security                  100.0
Department of Housing and Urban Development      100.0
Department of Interior                           100.0
Department of Justice                            100.0
Department of Labor                              100.0
Department of State                              100.0
Department of Transportation                     100.0
Department of Treasury                           100.0
Department of Veterans Affairs                   100.0
National Aeronautics and Space Administration    100.0
National Archives and Records Administration     100.0
Social Security Administration                   100.0
U.S. Agency for International Development        100.0

In [26]:
data[data["Department"] == 'Department of Health and Human Services'].groupby("Office", dropna=False).size()

Office
ACF Children's Bureau                                                                            1
AHRQ                                                                                             2
BARDA (CBRN & DRIVe)                                                                             2
BARDA (CBRN)                                                                                     4
BARDA (DRIVe)                                                                                    5
CBER/OBPV/DABRA                                                                                  3
CDER/Office of Generic Drugs                                                                     4
CDER/Office of New Drugs                                                                         1
CDER/Office of Pharmaceutical Quality (OPQ)                                                      1
CDER/Office of Strategic Programs (OSP)                                                          1
CDE

In [13]:
for col in data.columns:
    print(data.groupby(col, dropna=False).size().sort_values(ascending=False, na_position='first').head())

Use_Case_ID
DHS-0000-2023    1
HHS-0055-2023    1
HHS-0079-2023    1
HHS-0080-2023    1
HHS-0081-2023    1
dtype: int64
Department_Code
DOE    178
HHS    157
DOC     49
DHS     41
VA      40
dtype: int64
Agency
National Energy Technology Laboratory    127
NaN                                       76
NIH                                       47
FDA                                       44
USDA                                      39
dtype: int64
Office
NaN                                                   514
Centers for Medicare & Medicaid Services (CMS)         23
National Institutes of Health (NIH) NLM                14
NCTR                                                   12
CDER/Office of Surveillance and Epidemiology (OSE)     11
dtype: int64
Title
Machine Learning for Occupant Safety Research    2
Burn & Blast MCMs: Philips                       2
First Guess Excessive Rainfall Outlook           2
PyForecast                                       1
Privileged Material Identificat